src/y0/algorithm/identify/utils.py

```diff
@@ -0,0 +1,265 @@
+# -*- coding: utf-8 -*-
+
+"""Utilities for identification algorithms."""
+
+from __future__ import annotations
+
+from typing import Any, Iterable, Optional, Union
+
+import networkx as nx
+from ananke.graphs import ADMG
+
+from y0.dsl import (
+    CounterfactualVariable,
+    Expression,
+    P,
+    Probability,
+    Variable,
+    get_outcomes_and_treatments,
+)
+from y0.graph import NxMixedGraph
+from y0.mutate.canonicalize_expr import expr_equal
+
+__all__ = [
+    "Query",
+    "Identification",
+    "Fail",
+    "str_nodes_to_variable_nodes",
+]
+
+
+class Fail(Exception):
+    """Raised on failure of the identification algorithm."""
+
+
+class Query:
+    """An identification query."""
+
+    outcomes: set[Variable]
+    treatments: set[Variable]
+    conditions: set[Variable]
+
+    def __init__(
+        self,
+        outcomes: set[Variable],
+        treatments: set[Variable],
+        conditions: Optional[set[Variable]] = None,
+    ) -> None:
+        """Instantiate an identification.
+
+        :param outcomes: The outcomes in the query
+        :param treatments: The treatments in the query (e.g., counterfactual variables)
+        :param conditions: The conditions in the query (e.g., coming after the bar)
+        :raises TypeError: If any of the outcomes, treatements, or conditions are not vanilla variables
+        """
+        if not all(isinstance(v, Variable) for v in outcomes):
+            raise TypeError
+        elif any(isinstance(v, CounterfactualVariable) for v in outcomes):
+            raise TypeError
+        else:
+            self.outcomes = outcomes
+
+        if not all(isinstance(v, Variable) for v in treatments):
+            raise TypeError
+        elif any(isinstance(v, CounterfactualVariable) for v in treatments):
+            raise TypeError
+        else:
+            self.treatments = treatments
+
+        if conditions is None:
+            self.conditions = set()
+        elif not all(isinstance(v, Variable) for v in conditions):
+            raise TypeError
+        elif any(isinstance(v, CounterfactualVariable) for v in conditions):
+            raise TypeError
+        else:
+            self.conditions = conditions
+
+    @classmethod
+    def from_expression(
+        cls,
+        query: Probability,
+    ) -> Query:
+        """Instantiate an identification.
+
+        :param query: The query probability expression
+        :returns: An identification tuple
+        """
+        outcomes, treatments = get_outcomes_and_treatments(query=query)
+        outcomes = {Variable(v.name) for v in outcomes}  # clean counterfactuals
```

Comment:
@djinnome should outcomes be all of the parents + children of the distribution? or just the children

Reply:
`outcomes` should be children.   `conditions` should be parents

Reply:
Fixed in ce8a659